### PR TITLE
Tracing: Drawer view for span details

### PR DESCRIPTION
**Pull request by joey-grafana**
**What is this feature?**

This modifies the trace view to use a drawer for each spans details. Within this drawer, tabs have been added for each of the span details main sections i.e. tag, events, references. All other information from within the previously expanded span details has been retained and incorporated into the new design.

Screenshot 2023-10-23 at 13 09 08

**Why do we need this feature?**

Makes it easier to keep context of your span details as you continue to search your trace. Arranges all of the spans details into tabs for easy navigation.

**Who is this feature for?**

Users of the trace view.

**Special notes for your reviewer:**

Please use the `traceViewDrawer` feature toggle to enable.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/TraceView/components/DetailsPanel/AccordianLogs.tsx
Comment (joey-grafana) [inline public/app/features/explore/TraceView/components/DetailsPanel/AccordianLogs.tsx:47]: Mostly copy of original `AccordianLogs` component with design tweaks to suit card layout with tabs.

File: public/app/features/explore/TraceView/components/DetailsPanel/AccordianReferences.tsx
Comment (joey-grafana) [inline public/app/features/explore/TraceView/components/DetailsPanel/AccordianReferences.tsx:94]: Again this is mostly a copy of original `References` component with design tweaks to suit card layout with tabs.

File: public/app/features/explore/TraceView/useDetailState.ts
Comment (joey-grafana) [inline public/app/features/explore/TraceView/useDetailState.ts:36]: Will essentially be removed (with other checks for config.featureToggles.traceViewDrawer being enabled) when toggle is removed.

**Comment by ifrost** [comment]
There are a few issues with the current approach we see from Explore POV. The drawer looks like an Explore drawer, it's detached from the Trace View making UI confusing:

- it clashes with existing drawers (Query History, Inspector), 
- the scrolling context is not clear, 
- it doesn't work nicely with the Content Outline (especially when resized),
- it may overlap other visualizations 
- there's also a bug where resizing the drawer resizes the content of Explore

What is more, it'd be nice if Trace View at some point was extracted to a separate panel plugin so it could be used by application plugins. With this approach, it may be much harder to reuse it.

We think that it'd be better to have the drawer embedded inside the Trace View. Has this been considered?

Was this reviewed by UX? cc @diegoadams

**Comment by diegoadams** [comment]
@Elfo404 recorded a good video for me. I agree with all that is mentioned above, I think we may need some more exploration on this, some initial thoughts or alternative suggestions:

- Inline Expansion
Instead of a detached drawer, maybe an inline expansion beneath the selected trace. When a user clicks on a trace, the detailed information could slide down and appear directly below it, pushing other traces down. This keeps context and ensures users understand which trace details they're viewing.

- Sidebar Panel:
THis could push the main content slightly to make room for the trace details. This keeps the trace details constantly visible while the user interacts with other parts of the UI. But I worry it may interfere with the Content Outline or Amy's docked nav.

**Comment by joey-grafana** [comment]
Thank you all for the detailed feedback. It sounds like embedding the drawer inside the trace view would be the way to go. I've gone ahead and made changes for that 👍 Note we currently use inline expansion but the context of the span is essentially lost when you scroll away from that expanded span.

**Comment by joey-grafana** [comment]
@ifrost we're good to go for a re-review if you can take a look! (LMK if I should tag the team of do something else to reach ye)
@adrapereira 3 of 4 suggestions addressed. WRT stickying the header, feels like this might be trouble on smaller screens and with the multiple span details we will implement soon (so, tabs within tabs), there will be a lot of info on the screen and we may not actually want too much data/movement/sticky on screen. But leave it with me, I will keep an eye on this and possibly make the sticky change when we have added more functionality to the drawer.

**Comment by ifrost** [comment]
> @ifrost we're good to go for a re-review if you can take a look! (LMK if I should tag the team of do something else to reach ye) 

@grafana/explore-squad, also @diegoadams could you have a look at this?

One thing - could you create a separate drawer component instead of adding `className` to override ExploreDrawer? It looks a bit hacky and will be confusing in the future (e.g. that passing `className` overrides the width). ExploreDrawer is used in slightly different context and has some Explore-specific magic numbers for sizing and z-indices. We may need to change the component neat future, especially because:

- We plan to work on query history improvements (query history is in a drawer)
- We know about a few issues Explore drawers have with Content Outline
- There are issues with the inspector not showing correct state in split mode

We need to be able to make any changes safely without accidentally breaking TraceView. It will also allow you to adjust it easier in the future and also make it easier to externalize the plugin.

What's the rollout plan? Are you doing A/B testing or just rolling it out gradually?

File: public/app/features/explore/TraceView/components/DetailsPanel/DetailsPanel.tsx
Comment (harisrozajac) [inline public/app/features/explore/TraceView/components/DetailsPanel/DetailsPanel.tsx:270]: can we update this one to use an object, like this: 
```suggestion
  tab: css({
    label: 'DetailsPanelTab',
    padding: '0.5rem 1rem',
    '.json-markup': {
      lineHeight: '17px',
      fontSize: '13px',
      fontFamily: 'monospace',
      whiteSpace: 'pre-wrap',
    },
    '.json-markup-key': {
      fontWeight: 'bold',
    },
    '.json-markup-bool': {
      color: autoColor(theme, 'firebrick'),
    },
    '.json-markup-string': {
      color: autoColor(theme, 'teal'),
    },
    '.json-markup-null': {
      color: autoColor(theme, 'teal'),
    },
    '.json-markup-number': {
      color: autoColor(theme, 'blue', 'black'),
    },
  }),
```
I know that drawer can't be updated to use object because emotion doesn't support using `!important` in an object, but I thought we could at least update this one and reduce the number of betterer warnings

**Comment by harisrozajac** [comment]
I'm seeing a horizontal scrollbar in the drawer, although the content fits:
Screenshot 2023-11-01 at 8 40 49 AM

is this intended?

**Comment by Elfo404** [comment]
@ifrost Aside from everything technical i still don't understand why we are using a drawer in the first place. the drawer itself does not go above the trace but it basically creates a vertical split view.

using the drawer component and adding padding to the "top view" seems like an hack for something that should be instead done with a split view natively.

**Comment by ifrost** [comment]
> @ifrost Aside from everything technical i still don't understand why we are using a drawer in the first place. the drawer itself does not go above the trace but it basically creates a vertical split view.
> using the drawer component and adding padding to the "top view" seems like an hack for something that should be instead done with a split view natively.

cc @joey-grafana. I guess that would be even simpler to implement.

**Comment by joey-grafana** [comment]
@ifrost created a new component for the drawer. It's behind a feature toggle and we will progressively enable it in cloud.
@harisrozajac thanks, updated most in last commit!
@Elfo404 we want to view the trace and the span details together. We need to be able to see the entire trace even when the drawer is open. Are you suggesting a horizontal split view like in explore? That could work, but users would then see 3 split views in explore as they work with traces.

**Comment by Elfo404** [comment]
exactly, what is in this PR is not actually a drawer as it doesn't sit on top of the trace. if you don't want the bottom half to overlap the top one then my opinion is that we shouldn't use a drawer in the first place as that is its intended usage.

**Comment by joey-grafana** [comment]
@harisrozajac sizing fixed.

Not sure what you are suggesting here. The goal is to allow the span details to be viewed in a fixed place while also allowing users to scroll their trace and search for other spans. Currently, they can open span details but as soon as they scroll up or down they loose their trace and also the context. So if we're not supposed to use a drawer and a split view is not applicable what do we use? Do we really need to use something else?

**Comment by diegoadams** [comment]
Hey @joey-grafana , is there anyway you could please do a recording for me with the new updates so I can see what's going on, thanks/.

**Comment by joey-grafana** [comment]
Sure, this is a demo


https://github.com/grafana/grafana/assets/90795735/02a631c6-cbf1-4e3f-a079-4a750ecf0a9d

**Comment by ifrost** [comment]
> Not sure what you are suggesting here. The goal is to allow the span details to be viewed in a fixed place while also allowing users to scroll their trace and search for other spans. Currently, they can open span details but as soon as they scroll up or down they loose their trace and also the context. So if we're not supposed to use a drawer and a split view is not applicable what do we use? Do we really need to use something else?

Thanks @joey-grafana - changes to Explore are now simple after your update so I'm gonna approve it from Explore side so it's not blocking you 🙌  You're the code owner of the component so you can make the final call on what to use.

However, I think what @Elfo404 saying makes sense:

> the drawer itself does not go above the trace but it basically creates a vertical split view.
> what is in this PR is not actually a drawer as it doesn't sit on top of the trace.

It looks like it's a drawer, though it doesn't behave like any other Grafana drawers. Drawers "overlay" the content. Opening a drawer doesn't change the behavior/size/scrolling of what's behind. In case of the trace view drawer it's different - once the drawer is opened the content that is behind can be scrolled in a way that shows up above the drawer, which is not possible with "classic drawers". 

And I totally understand the reasoning behind such behavior. What you implemented makes sense - it should be always possible to scroll to any span 👍 having "a classic drawer" wouldn't make sense here. What you created is a very interesting UX pattern, but it's something new and could be confusing so I wonder what's the UX feedback on it? I'd also double check if such content is accessible.

Since it doesn't behave like any other drawer we advised it may not be the best option. I think what @Elfo404 suggested is creating a top-down split inside the trace view (not related to Explore's split view).  I guess after the split scrolling could also be adjusted in the top part so the selected span is always visible. This is not possible the moment - after selecting the last span, the user needs to scroll down to see it again in the tree view. It's up to you but we just wanted to share our concerns that something more predictable could be used here.

One more thing I noticed - it doesn't work well when filters are applied and prev/next buttons are used

https://github.com/grafana/grafana/assets/745532/9c004486-326a-49e2-9506-6fffa1ee48f7

**Review by ifrost — APPROVED** [review]
Explore code changes LGTM 👍

**Comment by joey-grafana** [comment]
No problem @ifrost, appreciate the input. I would not have merged until I have taken care of the feedback here / from what Diego suggested in the Slack thread :)

**Comment by joey-grafana** [comment]
@ifrost @Elfo404 @diegoadams thank you for your feedback, I have thought about the approach and have vastly simplified the implementation.

The benefit of this approach is that it does not need to use a Drawer or Split View. One of the problems with either Drawer/Split View is that it does not work well with small screens or dashboards where we do not have much screen space. 
Now, span details work as they did before, but also include a `Pin to details panel` button. On click, the details panel is added to the bottom of the Trace View. The benefit here is that we are not consuming more screen space which is already limited but with the added benefit that users can pin multiple spans and view them side by side for their comparison as discussed in our [design doc](https://docs.google.com/document/d/1zWJpVyfA1wAuUTexGyfhY2XRD88zBBi6HUnR7pZVKBg/edit#heading=h.52jbl6ih2hhs).

https://github.com/grafana/grafana/assets/90795735/4ec46ab8-8b89-4422-b0d8-3084371f7ba5

**Pull request closed without merging**
